spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1300,6 +1300,82 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
 the `getDateHeader` method cannot translate the header to a `Date`
 object, an `IllegalArgumentException` is thrown.
 
+=== Request URI Processing
+The process described here adapts and extends the URI canonicalization process described in [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) to create a standard Servlet URI path canonicalization process that ensures that URIs can be mapped to Servlets, Filters and security constraints in an unambiguous manner. It is also intended to provide information to reverse proxy implementations so they are aware of how requests they pass to servlet containers will be processed.
+
+Servlet containers may implement the standard Servlet URI path canonicalization in any manner they see fit as long as the end result is identical to the end result of the process described here. Servlet containers may provide container specific configuration options to vary the standard canonicalization process. Any such variations may have security implications and both Servlet container implementors and users are advised to be sure that they understand the implications of any such container specific canonicalization options.
+
+==== URI Transport
+===== HTTP/1.1
+The URI is extracted from the `request-target` as defined by [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1). URIs in `origin-form` or `asterisk-form` are passed unchanged to stage 2. URIs in `absolute-form` have the protocol and authority removed to convert them to `origin-form` and are then passed to stage 2. URIs in `authority-form` are outside of the scope of this specification.
+
+===== HTTP/2
+The URI is the `:path` pseudo header as defined by [RFC 7540](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3) and is passed unchanged to stage 2.
+
+===== Other protocols
+Containers may support other protocols. Containers should extract an appropriate URI for the request from the protocol and pass it to stage 2.
+
+==== Separation of path and query
+The URI is split by the first occurrence of any '?' character to path and query.  The query is preserved for later handling and the following steps applied to the path.
+
+==== Discard fragment
+A fragment in the path is indicated by the first occurrence of a `\#` character.  Any `#` character and following fragment is removed from the path and discarded.
+
+==== Decoding of non-special characters
+Characters other than `/`, `;` and `%` that are encoded in `%nn` form are decoded and the resulting octet sequences is treated as UTF-8 and converted to a character sequence.
```

Comment:
';' is not the only character that can be used for path segments.

You cannot avoid two separate passes for %nn decoding if you want to be RFC 3986 compliant w.r.t path parameters and to able to support the use of sub-delims in resource paths.

I do accept that the original text (decode `unreserved`) doesn't cover UTF-8 sequences representing code points U+0080 and above.

I think the character we do care about are `sub-delims` (since they can all be used for path parameters), `%2F` (unless it has already been rejected which it will have been by default) and %25 (since % is the escape). 

How about:

>decode (as UTF-8) %nn sequences apart from those representing `sub-delims` and `%`.

Reply:
But we don't act on any of the other sub-delims, so there is no need to keep them encoded only to decode them afterwards.    We will have already processed the special meanings of `:`, `?` and '#` by the time we get to this step.  The only other characters we parse specially after this point are `;` and `%`, so I think they are the only ones we need to leave encoded (so if anything I think we can drop `\` from here).

If we don't decode chars like `&` or `=` here, then we will just decode them later and the results we be the same. 

Hmmm perhaps `=` might be a little bit special for the processing of path params.   Actually I have an idea for an alternate formulation of the algorithm that might be a lot clearer (but perhaps further away from how it might actually be implemented).  Standby ....

Reply:
New proposal below.  I'll resolve this discussion now for clarity and we can discuss on the other proposal